src/device_trezor/trezor/protocol.cpp

```diff
@@ -0,0 +1,896 @@
+
+      } else {
+        throw std::invalid_argument("Unknown rsig type");
+      }
+    }
+  }
+
+  void Signer::compute_integrated_indices(TsxData * tsx_data){
+    if (m_aux_data == nullptr || m_aux_data->tx_recipients.empty()){
+      return;
+    }
+
+    auto & chg = tsx_data->change_dts();
+    std::string change_hash = hash_addr(&chg.addr(), chg.amount(), chg.is_subaddress());
+
+    std::vector<uint32_t> integrated_indices;
+    std::set<std::string> integrated_hashes;
+    for (auto & cur : m_aux_data->tx_recipients){
+      if (!cur.has_payment_id){
+        continue;
+      }
+      integrated_hashes.emplace(hash_addr(&cur.address.m_spend_public_key, &cur.address.m_view_public_key));
+    }
+
+    ssize_t idx = -1;
+    for (auto & cur : tsx_data->outputs()){
+      idx += 1;
+
+      std::string c_hash = hash_addr(&cur.addr(), cur.amount(), cur.is_subaddress());
+      if (c_hash == change_hash || cur.is_subaddress()){
+        continue;
+      }
+
+      c_hash = hash_addr(&cur.addr());
+      if (integrated_hashes.find(c_hash) != integrated_hashes.end()){
+        integrated_indices.push_back((uint32_t)idx);
+      }
+    }
+
+    if (!integrated_indices.empty()){
+      assign_to_repeatable(tsx_data->mutable_integrated_indices(), integrated_indices.begin(), integrated_indices.end());
+    }
+  }
+
+  std::shared_ptr<messages::monero::MoneroTransactionInitRequest> Signer::step_init(){
+    // extract payment ID from construction data
+    auto & tsx_data = m_ct.tsx_data;
+    auto & tx = cur_tx();
+
+    m_ct.tx.version = 2;
+    m_ct.tx.unlock_time = tx.unlock_time;
+
+    tsx_data.set_version(1);
+    tsx_data.set_unlock_time(tx.unlock_time);
+    tsx_data.set_num_inputs(static_cast<google::protobuf::uint32>(tx.sources.size()));
+    tsx_data.set_mixin(static_cast<google::protobuf::uint32>(tx.sources[0].outputs.size()));
+    tsx_data.set_account(tx.subaddr_account);
+    assign_to_repeatable(tsx_data.mutable_minor_indices(), tx.subaddr_indices.begin(), tx.subaddr_indices.end());
+
+    // Rsig decision
+    auto rsig_data = tsx_data.mutable_rsig_data();
+    m_ct.rsig_type = get_rsig_type(tx.use_bulletproofs, tx.splitted_dsts.size());
+    rsig_data->set_rsig_type(m_ct.rsig_type);
+
+    generate_rsig_batch_sizes(m_ct.grouping_vct, m_ct.rsig_type, tx.splitted_dsts.size());
+    assign_to_repeatable(rsig_data->mutable_grouping(), m_ct.grouping_vct.begin(), m_ct.grouping_vct.end());
+
+    translate_dst_entry(tsx_data.mutable_change_dts(), &(tx.change_dts));
+    for(auto & cur : tx.splitted_dsts){
+      auto dst = tsx_data.mutable_outputs()->Add();
+      translate_dst_entry(dst, &cur);
+    }
+
+    compute_integrated_indices(&tsx_data);
+
+    int64_t fee = 0;
+    for(auto & cur_in : tx.sources){
+      fee += cur_in.amount;
+    }
+    for(auto & cur_out : tx.splitted_dsts){
+      fee -= cur_out.amount;
+    }
+    if (fee < 0){
+      throw std::invalid_argument("Fee cannot be negative");
+    }
```

Comment:
That looks possibly integer overflowy, though one would have to have a LOT of monero to pull that off.

Reply:
Hmm the idea was just to add the sanity check for non-negative fee just to be sure as it is cast to unsigned type below. It should not be possible to overflow it on running networks (main, test, stage), I think.